sympy/physics/mechanics/joint.py

```diff
@@ -0,0 +1,591 @@
+    This is the base class for all specific joints and holds all common methods
+    acting as an interface for all joints. Custom joint can also be created
+    by creating a subclass of this class and overriding 'apply_joint()'
+    method to add the dynamics of the joint.
+
+    Parameters
+    ----------
+    name: String
+        Name of the joint which makes it unique. Should be different from other
+        joints.
+    parent: Body
+        Instance of Body class serving as the parent in the joint.
+    child: Body
+        Instance of Body class serving as the child in the joint.
+    parent_point_pos: 3 Tuple, Optional
+        Defines the joint's point where the parent will be connected to child.
+        3 Tuple defines the values of x, y and z directions w.r.t parent's
+        frame. If it is not supplied, center of mass is added as default.
+    child_point_pos: 3 Tuple, Optional
+        Defines the joint's point where the child will be connected to parent.
+        3 Tuple defines the values of x, y and z directions w.r.t child's frame.
+        If it is not supplied, center of mass is added as default.
+    parent_axis: Vector, Optional
+        Defines the orientation as a vector which must be aligned with child's
+        axis before adding joint. If it is not passed, default is x axis in
+        parent's frame.
+    child_axis: Vector, Optional
+        Defines the orientation as a vector which must be aligned with parent's
+        axis before adding joint. If it is not passed, default is x axis in
+        child's frame.
+    """
+
+    def __init__(self, name, parent, child, parent_point_pos=None,
+                 child_point_pos=None, parent_axis=None, child_axis=None):
+        self.name = name
+        self.parent = parent
+        self.child = child
+        self.coordinates = []
+        self.speeds = []
+        self.kds = []
+
+        if parent_point_pos is None:
+            parent_point_pos = (0, 0, 0)  # Parent's Center of mass
+
+        if child_point_pos is None:
+            child_point_pos = (0, 0, 0)  # Child's Center of mass
+
+        self.parent_joint_vector = self._convert_tuple_to_vector(
+            parent.frame, parent_point_pos)
+        self.child_joint_vector = self._convert_tuple_to_vector(
+            child.frame, child_point_pos)
+
+        if parent_axis is None:
+            self.parent_axis = parent.frame.x
+        else:
+            self.parent_axis = self._convert_tuple_to_vector(parent.frame,
+                                                             parent_axis)
+        if child_axis is None:
+            self.child_axis = child.frame.x
+        else:
+            self.child_axis = self._convert_tuple_to_vector(child.frame,
+                                                            child_axis)
+        self._locate_joint_point()
+        self.apply_joint()
+
+    def _convert_tuple_to_vector(self, frame, pos_tuple):
```

Comment:
I think you could move this to the functions module. It could even be public if you want.